### PR TITLE
Simplifying topo implementations.

### PR DESCRIPTION
**Pull request by alainjobart**
Now that all implementations are based on the topo.Backend sub-interface, this is possible.

BUG=30274258

### CODE REVIEW

**Comment by alainjobart** [comment]
@sougou ready for review. Probably one commit at a time is easier.

(will create an internal CL for it too)

**Comment by alainjobart** [comment]
And this: +938 −3,855 yes!

File: go/vt/topo/test/vschema.go [hunk 1]
```diff
@@ -48,69 +45,21 @@ func checkVSchema(t *testing.T, ts topo.Impl) {
...
-				ColumnVindexes: []*vschemapb.ColumnVindex{
-					{
-						Column: "c1",
-						Name:   "stfu1",
-					}, {
-						Column: "c2",
-						Name:   "stln1",
-					},
-				},
-			},
```
Comment (pivanof) [inline go/vt/topo/test/vschema.go]: Why did you remove the vschema definition here?

File: go/vt/vtctld/explorer.go [hunk 2]
```diff
@@ -21,42 +21,140 @@ import (
...
+// FIXME(alainjobart) GetKnownCells is only on topo.Impl at the moment.
+// Soon, when all topo implementations use the 'cells' subdirectory,
+// then we can use topo.Backend, as intended.
+type backendExplorer struct {
+	// backend is of type topo.Impl now, but will switch to topo.Backend.
+	backend topo.Impl
+}
+
+// Result is what the backendExplorer returns. It represents one directory node.
+// It is exported to the JSON encoder can print it.
```
Comment (pivanof) [inline go/vt/vtctld/explorer.go]: This sentence has two verbs.

File: go/vt/schemamanager/schemamanager_test.go [hunk 3]
```diff
@@ -282,6 +284,7 @@ func (client *fakeTabletManagerClient) ExecuteFetchAsDba(ctx context.Context, ta
 	return client.TabletManagerClient.ExecuteFetchAsDba(ctx, tablet, usePool, query, maxRows, disableBinlogs, reloadSchema)
 }
 
+// FIXME(alainjobart) replace with a memorytopo.
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/schemamanager/schemamanager_test.go]: I thought this would be too much for a single PR, but I guess it's huge already.

Pushed a commit with a fix for this, working on the rest now.

File: go/vt/topo/helpers/copy.go [hunk 4]
```diff
@@ -80,7 +80,9 @@ func CopyKeyspaces(ctx context.Context, fromTS, toTS topo.Impl) {
 
 // CopyShards will create the shards in the destination topo.
 func CopyShards(ctx context.Context, fromTS, toTS topo.Impl) {
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/helpers/copy.go]: Done.

File: go/vt/topo/replication.go [hunk 5]
```diff
@@ -157,3 +160,86 @@ func FixShardReplication(ctx context.Context, ts Server, logger logutil.Logger,
 	logger.Infof("All entries in replication graph are valid")
 	return nil
 }
+
+// UpdateShardReplicationFields updates the fields inside a topo.ShardReplication object.
+func (ts Server) UpdateShardReplicationFields(ctx context.Context, cell, keyspace, shard string, update func(*topodatapb.ShardReplication) error) error {
+	nodePath := path.Join(KeyspacesPath, keyspace, ShardsPath, shard, ShardReplicationFile)
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/replication.go]: I don't think we need to do that any more, it was a leftover from old code.

File: go/vt/topo/srv_keyspace.go [hunk 6]
```diff
@@ -92,3 +92,47 @@ func (ts Server) WatchSrvKeyspace(ctx context.Context, cell, keyspace string) (*
...
+	case ErrNoNode:
+		return nil, nil
+	default:
+		return nil, err
+	}
+}
+
+// UpdateSrvKeyspace saves a new SrvKeyspace. It is a blind write.
+func (ts Server) UpdateSrvKeyspace(ctx context.Context, cell, keyspace string, srvKeyspace *topodatapb.SrvKeyspace) error {
+	nodePath := srvKeyspaceFileName(keyspace)
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/srv_keyspace.go]: Done.

File: go/vt/topo/test/keyspace.go [hunk 7]
```diff
@@ -91,44 +91,30 @@ func checkKeyspace(t *testing.T, ts topo.Impl) {
 		t.Errorf("GetKeyspaces: want %v, got %v", []string{"test_keyspace", "test_keyspace2"}, keyspaces)
 	}
 
-	// re-read and update.
-	storedK, storedVersion, err := ts.GetKeyspace(ctx, "test_keyspace2")
+	// Re-read and update. Have to lock it.
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/test/keyspace.go]: UpdateKeyspace checks for the lock. Updated comment.

File: go/vt/topo/test/keyspace.go [hunk 8]
```diff
@@ -91,44 +91,30 @@ func checkKeyspace(t *testing.T, ts topo.Impl) {
...
 	if err != nil {
+		t.Fatalf("LockKeyspace: %v", err)
+	}
+	if err := ts.UpdateKeyspace(lockCtx, storedKI); err != nil {
 		t.Fatalf("UpdateKeyspace: %v", err)
 	}
-
-	// unconditional update
-	storedK.ShardingColumnType = topodatapb.KeyspaceIdType_BYTES
-	_, err = ts.UpdateKeyspace(ctx, "test_keyspace2", storedK, -1)
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/test/keyspace.go]: We don't do keyspace unconditional updates, so I removed the test case for it. The new API doesn't surface version=nil here.

File: go/vt/topo/test/lock.go [hunk 9]
```diff
@@ -34,7 +34,9 @@ var timeUntilLockIsTaken = 10 * time.Millisecond
 // checkKeyspaceLock checks we can take a keyspace lock as expected.
 func checkKeyspaceLock(t *testing.T, ts topo.Impl) {
 	ctx := context.Background()
-	if err := ts.CreateKeyspace(ctx, "test_keyspace", &topodatapb.Keyspace{}); err != nil {
+	// FIXME(alainjobart) when everywhere, just pass topo.Server as ts.
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/test/lock.go]: Next on my list is to push the Lock API down to the topo.Backend interface. Will be a different  PR though, this one is big already. But coming up next.

File: go/vt/topo/test/vschema.go (hunk 1 above)
Comment (alainjobart) [reply go/vt/topo/test/vschema.go]: The new higher level API checks it is correct, and this one wasn't.

File: go/vt/vtctld/explorer.go (hunk 2 above)
Comment (alainjobart) [reply go/vt/vtctld/explorer.go]: Fixed.

**Comment by alainjobart** [comment]
And thanks @pivanof for the review, this is a lot of cleanup. 👍

File: go/vt/topo/replication.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pivanof) [reply go/vt/topo/replication.go]: You mean now all the code everywhere uses only shard names in lower case and there is no possibility to pass shard names through a flag (or read it from topology)?

File: go/vt/topo/test/replication.go [hunk 10]
```diff
@@ -27,9 +27,10 @@ import (
 
 // checkShardReplication tests ShardReplication objects
 func checkShardReplication(t *testing.T, ts topo.Impl) {
```
Comment (pivanof) [inline go/vt/topo/test/replication.go]: It feels like this function needs to accept Server instead of Impl as well.

File: go/vt/schemamanager/tablet_executor_test.go [hunk 11]
```diff
@@ -48,14 +52,26 @@ func TestTabletExecutorOpen(t *testing.T) {
...
+			Cell: "test_cell",
+			Uid:  1,
+		},
+		Keyspace: "test_keyspace",
+		Shard:    "0",
+		Type:     topodatapb.TabletType_REPLICA,
+	}
+	// This will create the Keyspace, Shard and Tablet record.
+	// Since this is a replica tablet, the Shard will have no master.
+	if err := wr.InitTablet(ctx, tablet /*allowMasterOverride=*/, false /*createShardAndKeyspace=*/, true /*allowUpdate*/, false); err != nil {
```
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/schemamanager/tablet_executor_test.go]: I always get hit by this, gofmt has different format that C. Fixed.

File: go/vt/topo/replication.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/replication.go]: You can pass shard names through a flag, but if you use the wrong case, it will just fail (with unknown shard error). Same as for keyspaces, if you use the wrong case, it will fail. That's just more consistent.

But in most cases, we go through ValidateShard (see above) which already does the ToLower. So all storage is already lower case.

File: go/vt/topo/test/replication.go (hunk 10 above)
Comment (alainjobart) [reply go/vt/topo/test/replication.go]: Done.

**Comment by alainjobart** [comment]
@pivanof addressed latest comments, with commit.

One comment was lost in there, but I saw it somewhere:
>> Should actually NewTee accept Server instead of Impl and return Server as well?

I think so, but I also have on my list to refactor some of that code, so I'll do it then, if you don't mind. Right now, the tee is at the Impl level, it should either be at Backend, or Server, as Impl is going to disappear. So this code will need to change at some point again.

File: go/vt/topo/replication.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alainjobart) [reply go/vt/topo/replication.go]: A bit more on this: all code path that alter the topo go through ValidateShard, or TabletComplete, which will both fix the shard name to lower case for keyranges. So we really guarantee everything is lower case for keyranges (like a0-b0). For non-keyranges, we will just keep the case, for all topo implementations alike.
